grand_challenge_forge/partials/example-evaluation-method/example-evaluation-method{{cookiecutter._}}/evaluate.py.j2

```diff
@@ -44,13 +58,12 @@ def main():
     # Note that the jobs are not in any order!
     # We work that out from predictions.json
 
-    # Start a number of process workers, using multiprocessing
-    # The optimal number of workers ultimately depends on how many
-    # resources each process() would call upon
-    with Pool(processes=4) as pool:
-        metrics["results"] = pool.map(process, predictions)
+    # Use threaded workers to process the predictions more efficiently
+    with ThreadPoolExecutor(max_workers=MAX_WORKERS) as executor:
```

Comment:
Not sure that using a ThreadPoolExecutor is the correct decision here. Users would now need to ensure that `process` is thread safe. This also would be vulnerable to CPU locking. In addition, you would still need to do extra handling of errors.

It is probably better to use the simpler API of concurrent.futures though. But still, more work needs to be done. In our case - CPU bound, no need to share memory, initialisation time << runtime - the `ProcessPoolExecutor` would be more appropriate. Here is roughly how to do it with error handling thanks to ChatGPT:

```python
from concurrent.futures import ProcessPoolExecutor, as_completed

def process_wrapper(prediction):
    try:
        return process(prediction)
    except Exception as e:
        return f"Error: {str(e)}"

results = []
errors = []

with ProcessPoolExecutor(max_workers=4) as executor:
    future_to_prediction = {executor.submit(process_wrapper, prediction): prediction for prediction in predictions}
    
    for future in as_completed(future_to_prediction):
        prediction = future_to_prediction[future]
        try:
            result = future.result()
            if isinstance(result, str) and result.startswith("Error:"):
                errors.append(result)
            else:
                results.append(result)
        except Exception as e:
            errors.append(f"Unexpected error for prediction {prediction}: {str(e)}")

if errors:
    print(f"Errors occurred: {errors}")
    # Handle errors as appropriate for your application
else:
    metrics["results"] = results
```

Reply:
Updated the example code.

Reply:
Here is even more with task cancelling:

```python
from concurrent.futures import ProcessPoolExecutor, as_completed, CancelledError
import os

results = []
errors = []

def process(prediction):
    # Your original process function implementation
    ...

def shutdown_executor(executor, futures):
    for future in futures:
        future.cancel()
    executor.shutdown(wait=False)

with ProcessPoolExecutor(max_workers=4) as executor:
    futures = [executor.submit(process, prediction) for prediction in predictions]
    future_to_prediction = {future: prediction for future, prediction in zip(futures, predictions)}
    
    try:
        for future in as_completed(future_to_prediction):
            prediction = future_to_prediction[future]
            try:
                result = future.result()
                results.append(result)
            except Exception as e:
                if isinstance(e.__cause__, (BrokenProcessPool, OSError)):
                    # Handle unexpected termination (e.g., OOM killer)
                    errors.append(f"Process was killed or terminated unexpectedly for prediction {prediction}: {str(e)}")
                else:
                    # General exception handling
                    errors.append(f"Error for prediction {prediction}: {str(e)}")
                
                # Cancel all other futures and shutdown the executor
                shutdown_executor(executor, futures)
                break
    except CancelledError:
        print("A task was cancelled.")

if errors:
    print(f"Errors occurred: {errors}")
    # Handle errors as appropriate for your application
else:
    metrics["results"] = results
```

Reply:
IF you use a ProcessPoolExecutor, you also might need to deal with subprocesses that might be spawned by the processes you create in a special way. Otherwise, you will accumulate zombie processes that if you "shutdown" the workers or a subprocess crashes.

See for example: https://stackoverflow.com/questions/7171722/how-can-i-handle-sigchld

Containing this brilliant sentence:

> Explicitly setting the disposition of SIGCHLD to SIG_IGN causes any child process that subsequently terminates to be immediately removed from the system **instead of being converted into a zombie.**

The problem with zombies is not only that they eat brains, but that they occupy system resources such as PIDs and sockets and memory, which can cause the subprocess to crash again. If you then launch the entire thing inside a container that does not run an `init`-process AND your application is not implementing "zombie reaping" they will just linger around and accumulate which is the worst case to be in.

Reply:
I don't mind adding these handlers, but there is a strong drive to keep the examples simple. The evaluation runs within a container as such I'd expect any zombies to be killed when the container stops. Maybe that is incorrect?

There is an argument for non-thread save operations. So `ProcessPoolExecutor` is a good alt.

Using the `ProcessPoolExecutor` and its `map` function seem to have constructs for detecting process killing and collecting errors allready.

For instance, using this example:

```Python
from concurrent.futures import ProcessPoolExecutor
import time


if __name__ == '__main__':
    def example_function(arg):
        print(f"Processing {arg}")
        time.sleep(10)

    with ProcessPoolExecutor(max_workers=4) as executor:
       results = list(executor.map(example_function, range(10)))

    print(results)
    ```
    
And killing one of the processes with `kill -9` gives: `concurrent.futures.process.BrokenProcessPool: A process in the process pool was terminated abruptly while the future was running or pending.`

In addition, adding a `raise RuntimeError()` seems to be picked up nicely as well.

As such, I don't see much use in overly complicated additions. We can read it fine, but we can't expect every challenge org to be on the same level.

Reply:
We want to stop the processes as it saves us money. We know that some organisers are planning evaluations that last 24 hours. We need all the results to compute the metrics, so it is better to cancel all the subprocesses than to let them all run. If you want to hide the code put it in another file.

Reply:
> We want to stop the processes as it saves us money. We know that some organisers are planning evaluations that last 24 hours. We need all the results to compute the metrics, so it is better to cancel all the subprocesses than to let them all run. If you want to hide the code put it in another file.

Ah, I hadn't realized that the regular `map()` is not a fail-fast construct. Yeah, we'd want that.

Reply:
I adkes ChatGPT myself with a different prompt. The "seemingly correct way" to reap processes is something like this:

~~~~~~~~~.py
import os
import signal
import time
import psutil

def enumerate_and_reap_zombies():
    # Get the current process
    current_process = psutil.Process(os.getpid())
    # Iterate through all child processes
    for child in current_process.children(recursive=False):
        try:
            # Check if the child process is a zombie
            if child.status() == psutil.STATUS_ZOMBIE:
                print(f'Found zombie process: {child.pid}')
                # Reap the zombie process
                os.waitpid(child.pid, 0)
                print(f'Reaped zombie process: {child.pid}')
        except psutil.NoSuchProcess:
            continue
~~~~~~~~~

"Seemingly", because I just pasted some response by ChatGPT here. I ddi not test it but the construct seems sound. `psutil` is necessary only if you need to support cross-platform execution. If not, you can also enumerate the `/proc` data structure manually. Not sure what is preferred in this case?

Reply:
I first had a depedency `pebble` added, since their implementation seemed to work. After adding some tests I could more readily play around with it and found a construct that works without `pebble`.

In short, an executor is spawned in a seperate process. If an error occurs or a child is killed, ALL child processes of that process are killed in a take-no-prisoners manner.

Should be quite robust now.

Reply:
Hey ... ok... I took some time today to do a thorough review of my concerns, but _I cannot reproduce this issue_. This does not mean that I do not think that a zombie issue exists, but right now I cannot quite reconstruct how this would cause problems for us. Unless you are mass-spawning processes during the runtime of you application?! - I mean in the number of 100k+ subprocesses?

_also ... "zombie issue" lol_